File: GameEngine_PC/GameEngineOpenGL2_1/src/main/java/gameEngine/shaders/ShaderManager.java
package gameEngine.shaders;

import org.lwjgl.opengl.GL20;

import com.dferreira.commons.GLSLUtils;
import com.dferreira.commons.GLTransformation;
import com.dferreira.commons.LoadUtils;
import com.dferreira.commons.ShaderProgram;
import com.dferreira.commons.Vector3f;

/**
 * Generic shader manager with methods to load vertex shader and fragment shader
 * files as well as useful methods to bind arguments
 */
public abstract class ShaderManager {

	/**
	 * Directory where the shader are in the project
	 */
	protected static final String COMMON_PATH = "src/main/java/GameEngine/Shaders/";

	private ShaderProgram shaderProgram;

	/**
	 * Constructor of the program shader manager
	 * 
	 * @param vertexFile
	 *            The file with vertex description
	 * @param fragmentFile
	 *            The file with fragment description
	 */
	public ShaderManager(String vertexFile, String fragmentFile) {

		String vertexShaderSrc = LoadUtils.readTextFromRawResource(vertexFile);
		String fragShaderSrc = LoadUtils.readTextFromRawResource(fragmentFile);
		this.shaderProgram = GLSLUtils.loadProgram(vertexShaderSrc, fragShaderSrc);

		if (this.shaderProgram == null) {
			return;
		}
		bindAttributes();
		boolean linked = GLSLUtils.linkProgram(shaderProgram);
		if (linked) {
			getAllUniformLocations();
		}
	}

	/**
	 * Called to bind the attributes to the program shader
	 */
	protected abstract void bindAttributes();

	/**
	 * Called to ensure that all the shader managers get their uniform locations
	 */
	protected abstract void getAllUniformLocations();

	/**
	 * Bind one attribute
	 * 
	 * @param attributeIndex
	 *            Index of the attribute to bind
	 * @param variableName
	 *            Name of the attribute to bind
	 */
	protected void bindAttribute(int attributeIndex, String variableName) {
		GL20.glBindAttribLocation(shaderProgram.getProgramId(), attributeIndex, variableName);
	}

	/**
	 * Get the position of one uniform variable in the program shader
	 * 
	 * @param uniformName
	 *            the name of the uniform variable as appears in the shader code
	 * 
	 * @return the position of the uniform variable in program shader
	 */
	protected int getUniformLocation(String uniformName) {
		return GL20.glGetUniformLocation(shaderProgram.getProgramId(), uniformName);
	}

	/**
	 * Get the location of the attribute in the program shader
	 * 
	 * @param attributeName
	 *            the name of the attribute to get the position
	 * 
	 * @return The location of the attribute in the program shader
	 */
	protected int getAttributeLocation(String attributeName) {
		return GL20.glGetAttribLocation(shaderProgram.getProgramId(), attributeName);
	}

	/**
	 * Load a integer value to be used in the shader script
	 * 
	 * @param location
	 *            location of the shader variable in the script
	 * @param value
	 *            The value to load
	 */
	protected void loadInt(int location, int value) {
		GL20.glUniform1i(location, value);
	}

	/**
	 * Load a float value to be used in the shader script
	 * 
	 * @param location
	 *            location of the shader variable in the script
	 * @param value
	 *            value to load
	 */
	protected void loadFloat(int location, float value) {
		GL20.glUniform1f(location, value);
	}

	/**
	 * Load a vector to be used in the shader script
	 * 
	 * @param location
	 *            location of the shader variable in the script
	 * @param vector
	 *            The vector to load
	 */
	protected void loadVector(int location, Vector3f vector) {
		GL20.glUniform3f(location, vector.x, vector.y, vector.z);
	}

	/**
	 * Load a boolean value to be used in the shader script
	 * 
	 * @param location
	 *            The location of the shader variable in the script
	 * @param value
	 *            value to load
	 */
	protected void loadBoolean(int location, boolean value) {
		float toLoad = value ? 1 : 0;
		GL20.glUniform1f(location, toLoad);
	}

	/**
	 * Load a matrix to be used in the shader script
	 * 
	 * @param location
	 *            The location of the shader variable in the script
	 * @param matrix
	 *            Matrix to load
	 */
	protected void loadMatrix(int location, GLTransformation matrix) {
		GL20.glUniformMatrix4(location, false, matrix.getAsFloatBuffer());
	}

	/**
	 * Indicates that should start to use a certain program shader
	 */
	public void start() {
		if (shaderProgram != null) {
			GL20.glUseProgram(shaderProgram.getProgramId());
		}
	}

	/**
	 * Indicate that should not use a certain program no more
	 */
	public void stop() {
		GL20.glUseProgram(0);
	}

	/**
	 * A bit of memory management
	 */
	public void cleanUp() {
		this.stop();

		if (shaderProgram != null) {
			int vertexShaderID = shaderProgram.getVertexShaderId();
			int fragmentShaderID = shaderProgram.getFragmentShaderId();
			int programId = shaderProgram.getProgramId();

			GL20.glDetachShader(programId, vertexShaderID);
			GL20.glDetachShader(programId, fragmentShaderID);
			GL20.glDeleteShader(vertexShaderID);
			GL20.glDeleteShader(fragmentShaderID);
			GL20.glDeleteProgram(programId);
		}

	}
}

File: GameEngine_IOS/GameEngine/Shaders/ShaderManager.swift
import Foundation
import OpenGLES

/**
 *  Master class to handle shaders
 *
 */
open class ShaderManager : NSObject{
    
    open var shaderProgram : ShaderProgram!
    internal var uniforms : Array<Int>
    
    /**
     * Initialize of the program shader manager
     *
     * @param vertexFile   File with vertex description
     * @param fragmentFile File with fragment description
     * @param numberOfUniforms number of uniform locations of the shader program
     */
    public init(vertexFile : String, fragmentFile : String, numberOfUniforms : Int) {
        self.uniforms = Array<Int>(repeating: -1, count: numberOfUniforms)
        
        super.init()
        
        var vertShaderPathname, fragShaderPathname: String
        var vertexShaderSrc, fragShaderSrc : NSString?
        
        //Read the vertex file to one variable
        vertShaderPathname = Bundle.main.path(forResource: vertexFile, ofType: "vsh")!
        vertexShaderSrc = LoadUtils.readText(vertShaderPathname);
        
        //Read the fragment shader file to one variable
        fragShaderPathname = Bundle.main.path(forResource: fragmentFile, ofType: "fsh")!
        fragShaderSrc = LoadUtils.readText(fragShaderPathname);
        
        self.shaderProgram = GLSLUtils.loadProgram(vertexShaderSrc, fragShaderSrc: fragShaderSrc);
        
        if (shaderProgram == nil) {
            return;
        }
        
        self.bindAttributes();
        
        // Link program.
        
        if (GLSLUtils.linkProgram(shaderProgram)) {
            self.getAllUniformLocations();
        } else {
            NSLog("Failed to link program: %d", shaderProgram.programId);
            
            if (shaderProgram.vertexShaderId != 0) {
                glDeleteShader(shaderProgram.vertexShaderId);
                shaderProgram.vertexShaderId = 0;
            }
            if (shaderProgram.fragmentShaderId != 0) {
                glDeleteShader(shaderProgram.fragmentShaderId);
                shaderProgram.fragmentShaderId = 0;
            }
            if (shaderProgram.programId != 0) {
                glDeleteProgram(shaderProgram.programId);
                shaderProgram.programId = 0;
            }
            
        }
    }
    
    /*
     * Called to provide the attributes that are goin to get bound to the shader
     */
    internal func getAttributes() -> Dictionary<Int, String>! {
        NSException.raise(NSExceptionName(rawValue: "Invoked abstract method"), format: "Invoked abstract method", arguments:getVaList(["nil"]))
        return nil;
    }
    
    
    /**
     * Called to bind the attributes to the program shader
     */
    fileprivate func bindAttributes() {
        let attributes : Dictionary<Int, String>! = self.getAttributes();
        
        
        if(attributes != nil) {
            for (key, value) in attributes {
                self.bindAttribute(key, variableName: value);
            }
        }
    }
    
    /**
     * Called to provide the uniform locations that are goin to get bound to the shader
     */
    internal func getUniformLocations() -> Dictionary<NSInteger, String>! {
        NSException.raise(NSExceptionName(rawValue: "Invoked abstract method"), format: "Invoked abstract method", arguments:getVaList(["nil"]))
        return nil;
    }
    
    /**
     * Called to ensure that all the shader managers get their uniform locations
     */
    fileprivate func getAllUniformLocations() {
        
        let uniformsDic : Dictionary<NSInteger, String> = getUniformLocations();
        
        //Iteract over different elements to get the locations in the program shader
        for (key, value) in uniformsDic {
            uniforms[key] = getUniformLocation(value)
            if(uniforms[key] < 0) {
                NSLog("Problems getting %@'s location in the shader", value);
            }
        }
    }
    
    /**
     * Bind one attribute
     *
     * @param attributeIndex Index of the attribute to bind
     * @param variableName   Name of the attribute to bind
     */
    fileprivate func bindAttribute(_ attributeIndex : Int , variableName : String) {
        glBindAttribLocation(shaderProgram.programId, GLuint(attributeIndex), variableName);
    }
    
    /**
     * Get the position of one uniform variable in the program shader
     *
     * @param uniformName the name of the uniform variable as appears in the shader code
     * @return the position of the uniform variable in program shader
     */
    internal func  getUniformLocation(_ uniformName : String) -> Int {
        return Int(glGetUniformLocation(shaderProgram.programId, uniformName));
    }
    
    /**
     * Load a integer value to be used in the shader script
     *
     * @param location
     *            location of the shader variable in the script
     * @param value
     *            The value to load
     */
    internal func loadInt(_ location : Int, value : Int) {
        glUniform1i(GLint(location), Int32(value));
    }
    
    /**
     * Load a float value to be used in the shader script
     *
     * @param location location of the shader variable in the script
     * @param value    value to load
     */
    internal func loadFloat(_ location : Int, value : Float) {
        glUniform1f(GLint(location), value);
    }
    
    /**
     * Load a 2D vector to be used in the shader script
     *
     * @param location
     *            location of the shader variable in the script
     * @param vector
     *            The vector to load
     */
    internal func loadVector(_ location : Int, vector : Vector2f) {
        glUniform2f(GLint(location), vector.x, vector.y);
    }
    
    /**
     * Load a 3D vector to be used in the shader script
     *
     * @param location location of the shader variable in the script
     * @param vector The vector to load
     */
    internal func  loadVector(_ location : Int, vector : Vector3f) {
        glUniform3f(GLint(location), vector.x, vector.y, vector.z);
    }
    
    /**
     * Load a color RGB to be used in the shader script
     *
     * @param location
     *            location of the shader variable in the script
     * @param color
     *            The color to load
     */
    internal func loadColorRGB(_ location: Int,  color: ColorRGB) {
        glUniform3f(GLint(location), color.r, color.g, color.b);
    }
    
    /**
     * Load a color RGBA to be used in the shader script
     *
     * @param location
     *            location of the shader variable in the script
     * @param color
     *            The color to load
     */
    internal func loadColorRGBA(_ location : Int, color: ColorRGBA) {
        glUniform4f(GLint(location), color.r, color.g, color.b, color.a);
    }
    
    /**
     * Load a boolean value to be used in the shader script
     *
     * @param location The location of the shader variable in the script
     * @param value    value to load
     */
    internal func  loadBoolean(_ location : Int, value : Bool) {
        let toLoad : Float = value ? 1.0 : 0.0;
        glUniform1f(GLint(location), toLoad);
    }
    
    /**
     * Load a matrix to be used in the shader script
     *
     * @param location The location of the shader variable in the script
     * @param matrix   Matrix to load
     */
    open func loadMatrix (_ location : Int, matrix : GLTransformation) {
        glUniformMatrix4fv(GLint(location), 1, 0, matrix.getMatrix());
    }
    
    /**
     * Indicates that should start to use a certain program shader
     */
    open func start() {
        if (shaderProgram != nil) {
            glUseProgram(shaderProgram.programId);
        }
    }
    
    /**
     * Indicate that should not use a certain program no more
     */
    open func  stop() {
        glUseProgram(0);
    }
    
    /**
     * Clean the program shader from memory
     */
    deinit {
        if ((shaderProgram != nil) && (shaderProgram.programId != 0)) {
            glDeleteProgram(shaderProgram.programId);
            shaderProgram.programId = 0;
        }
        shaderProgram = nil;
    }
    
}
